### ISSUE TITLE
Uninitialized constant in rails after renaming class

### ISSUE THREAD

**Issue by Earlopain** [fxn/zeitwerk#208]
Hi there, I'm having issues after renaming a class. Sometimes I mistype the classname when creating new files. Renaming it to the correct one results in all kinds of issues.

An example: I create a file `a.rb` defining the class `B`. This obviously won't work and zeitwerk displays the proper error message to me, so I rename the class to `A` to be in line with the filename. Nothing works anymore and I have to restart the server. Even if I delete the file completly zeitwerk is not able to recover.

I tested this with ruby 3.1, rails 7.0.1 with zeitwerk 2.5.3, nothing additional like bootsnap for example. I have once again made a minimally reproducable repro for you to take a look at if you want: https://github.com/Earlopain/zeitwerk-repro/tree/repro-2

<details>
<summary>
Logs
</summary>

```
Initial boot:
Zeitwerk@rails.main: autoload set for ApplicationController, to be loaded from /demo/app/controllers/application_controller.rb
Zeitwerk@rails.main: autoload set for A, to be loaded from /demo/app/test/a.rb

First request:
Zeitwerk@rails.main: autoload for ApplicationController removed
Zeitwerk@rails.main: autoload for A removed
Zeitwerk@rails.main: autoload set for ApplicationController, to be loaded from /demo/app/controllers/application_controller.rb
Zeitwerk@rails.main: autoload set for A, to be loaded from /demo/app/test/a.rb
Zeitwerk@rails.main: constant ApplicationController loaded from file /demo/app/controllers/application_controller.rb

ActionView::Template::Error (expected file /demo/app/test/a.rb to define constant A, but didn't
        raise Zeitwerk::NameError.new("expected file #{file} to define constant #{cpath}, but didn't", cref.last)
        ^^^^^):
      1: <%= A %>

```

Everything up to here is what I am expecting. Now it's getting buggy though.

```
Renamed class `B` to `A` in `a.rb`
Zeitwerk@rails.main: ApplicationController unloaded
NameError (uninitialized constant A

            value = parent.const_get(cname)
                          ^^^^^^^^^^):
  
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:148:in `const_get'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:148:in `block (2 levels) in unload'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:146:in `each'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:146:in `block in unload'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:125:in `synchronize'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:125:in `unload'
zeitwerk (2.5.3) lib/zeitwerk/loader.rb:194:in `reload'

Second request:
ActionController::RoutingError (uninitialized constant ApplicationController

      Object.const_get(camel_cased_word)
            ^^^^^^^^^^

            raise MissingController.new(error.message, error.name)
            ^^^^^):
  
```

</details>

**Comment by fxn** [comment]
I've published 2.5.4 with the fix.

**Comment by Earlopain** [comment]
Thank you, much appreciated
